File: Tablut/src/it/unibo/ai/didattica/competition/tablut/client/TablutRandomClient.java
package it.unibo.ai.didattica.competition.tablut.client;

import java.io.IOException;
import java.net.UnknownHostException;
import java.util.ArrayList;
import java.util.List;
import java.util.Objects;
import java.util.Random;

import it.unibo.ai.didattica.competition.tablut.domain.*;
import it.unibo.ai.didattica.competition.tablut.domain.State.Turn;

/**
 * 
 * @author A. Piretti, Andrea Galassi
 *
 */
public class TablutRandomClient extends TablutClient {

	private final int game;

	public TablutRandomClient(String player, String name, int gameChosen, int timeout, String ipAddress) throws IOException {
		super(player, name, timeout, ipAddress);
		game = gameChosen;
	}
	
	public TablutRandomClient(String player, String name, int timeout, String ipAddress) throws IOException {
		this(player, name, 4, timeout, ipAddress);
	}
	
	public TablutRandomClient(String player, int timeout, String ipAddress) throws IOException {
		this(player, "random", 4, timeout, ipAddress);
	}

	public TablutRandomClient(String player) throws IOException {
		this(player, "random", 4, 60, "localhost");
	}


	public static void main(String[] args) throws IOException {
		int gametype = 4;
		String role = "";
		String name = "random";
		String ipAddress = "localhost";
		int timeout = 60;
		// TODO: change the behavior?
		if (args.length < 1) {
			System.out.println("You must specify which player you are (WHITE or BLACK)");
			System.exit(-1);
		} else {
			System.out.println(args[0]);
			role = (args[0]);
		}
		if (args.length == 2) {
			System.out.println(args[1]);
			timeout = Integer.parseInt(args[1]);
		}
		if (args.length == 3) {
			ipAddress = args[2];
		}
		System.out.println("Selected client: " + args[0]);

		TablutRandomClient client = new TablutRandomClient(role, name, gametype, timeout, ipAddress);
		client.run();
	}

	@Override
	public void run() {

		try {
			this.declareName();
		} catch (Exception e) {
			e.printStackTrace();
		}

		State state;

		Game rules = null;
		switch (this.game) {
			case 1 -> {
				state = new StateTablut();
				rules = new GameTablut();
			}
			case 2 -> {
				state = new StateTablut();
				rules = new GameModernTablut();
			}
			case 3 -> {
				state = new StateBrandub();
				rules = new GameTablut();
			}
			case 4 -> {
				state = new StateTablut();
				state.setTurn(Turn.WHITE);
				rules = new GameAshtonTablut(99, 0, "garbage", "fake", "fake");
				System.out.println("Ashton Tablut game");
			}
			default -> {
				System.out.println("Error in game selection");
				System.exit(4);
			}
		}

		List<int[]> pawns = new ArrayList<>();
		List<int[]> empty = new ArrayList<>();

		System.out.println("You are player " + this.getPlayer().toString() + "!");

		while (true) {
			try {
				this.read();
			} catch (IOException e1) {
				// TODO Auto-generated catch block
				e1.printStackTrace();
				System.exit(1);
			}
			System.out.println("Current state:");
			state = this.getCurrentState();
			System.out.println(state.toString());
			try {
				Thread.sleep(1000);
			} catch (InterruptedException e) {
				throw new RuntimeException(e);
			}

			if (this.getPlayer().equals(Turn.WHITE)) {
				// Mio turno
				if (this.getCurrentState().getTurn().equals(StateTablut.Turn.WHITE)) {
					int[] buf;
					for (int i = 0; i < state.getBoard().length; i++) {
						for (int j = 0; j < state.getBoard().length; j++) {
							if (state.getPawn(i, j).equalsPawn(State.Pawn.WHITE.toString())
									|| state.getPawn(i, j).equalsPawn(State.Pawn.KING.toString())) {
								buf = new int[2];
								buf[0] = i;
								buf[1] = j;
								pawns.add(buf);
							} else if (state.getPawn(i, j).equalsPawn(State.Pawn.EMPTY.toString())) {
								buf = new int[2];
								buf[0] = i;
								buf[1] = j;
								empty.add(buf);
							}
						}
					}

					int[] selected;

					boolean found = false;
					Action a = null;
					a = new Action("z0", "z0", Turn.WHITE);

					while (!found) {
						if (pawns.size() > 1) {
							selected = pawns.get(new Random().nextInt(pawns.size() - 1));
						} else {
							selected = pawns.get(0);
						}

						String from = this.getCurrentState().getBox(selected[0], selected[1]);

						selected = empty.get(new Random().nextInt(empty.size() - 1));
						String to = this.getCurrentState().getBox(selected[0], selected[1]);

						a = new Action(from, to, Turn.WHITE);


						try {
							rules.checkMove(state, a);
							found = true;
						} catch (Exception e) {
							throw new RuntimeException(e);
						}

					}

					System.out.println("Mossa scelta: " + Objects.requireNonNull(a));
					try {
						this.write(a);
					} catch (IOException e) {
						// TODO Auto-generated catch block
						e.printStackTrace();
					}
					pawns.clear();
					empty.clear();

				}
				// Turno dell'avversario
				else if (state.getTurn().equals(StateTablut.Turn.BLACK)) {
					System.out.println("Waiting for your opponent move... ");
				}
				// ho vinto
				else if (state.getTurn().equals(StateTablut.Turn.WHITEWIN)) {
					System.out.println("YOU WIN!");
					System.exit(0);
				}
				// ho perso
				else if (state.getTurn().equals(StateTablut.Turn.BLACKWIN)) {
					System.out.println("YOU LOSE!");
					System.exit(0);
				}
				// pareggio
				else if (state.getTurn().equals(StateTablut.Turn.DRAW)) {
					System.out.println("DRAW!");
					System.exit(0);
				}

			} else {

				// Mio turno
				if (this.getCurrentState().getTurn().equals(StateTablut.Turn.BLACK)) {
					int[] buf;
					for (int i = 0; i < state.getBoard().length; i++) {
						for (int j = 0; j < state.getBoard().length; j++) {
							if (state.getPawn(i, j).equalsPawn(State.Pawn.BLACK.toString())) {
								buf = new int[2];
								buf[0] = i;
								buf[1] = j;
								pawns.add(buf);
							} else if (state.getPawn(i, j).equalsPawn(State.Pawn.EMPTY.toString())) {
								buf = new int[2];
								buf[0] = i;
								buf[1] = j;
								empty.add(buf);
							}
						}
					}

					int[] selected;

					boolean found = false;
					Action a = null;
					a = new Action("z0", "z0", Turn.BLACK);

					while (!found) {
						selected = pawns.get(new Random().nextInt(pawns.size() - 1));
						String from = this.getCurrentState().getBox(selected[0], selected[1]);

						selected = empty.get(new Random().nextInt(empty.size() - 1));
						String to = this.getCurrentState().getBox(selected[0], selected[1]);

						a = new Action(from, to, Turn.BLACK);


						System.out.println("try: " + Objects.requireNonNull(a));
						try {
							rules.checkMove(state, a);
							found = true;
						} catch (Exception e) {
							throw new RuntimeException(e);
						}

					}

					System.out.println("Mossa scelta: " + a);
					try {
						this.write(a);
					} catch (IOException e) {
						// TODO Auto-generated catch block
						e.printStackTrace();
					}
					pawns.clear();
					empty.clear();

				}

				else if (state.getTurn().equals(StateTablut.Turn.WHITE)) {
					System.out.println("Waiting for your opponent move... ");
				} else if (state.getTurn().equals(StateTablut.Turn.WHITEWIN)) {
					System.out.println("YOU LOSE!");
					System.exit(0);
				} else if (state.getTurn().equals(StateTablut.Turn.BLACKWIN)) {
					System.out.println("YOU WIN!");
					System.exit(0);
				} else if (state.getTurn().equals(StateTablut.Turn.DRAW)) {
					System.out.println("DRAW!");
					System.exit(0);
				}

			}
		}

	}
}

File: Tablut/src/it/unibo/ai/didattica/competition/tablut/client/TablutRandomClient.kt
package it.unibo.ai.didattica.competition.tablut.client

import it.unibo.ai.didattica.competition.tablut.domain.*
import it.unibo.ai.didattica.competition.tablut.domain.State.Pawn
import it.unibo.ai.didattica.competition.tablut.domain.State.Turn
import java.io.IOException
import java.util.*

/**
 *
 * @author A. Piretti, Andrea Galassi
 */
class TablutRandomClient @JvmOverloads constructor(
    player: String,
    name: String = "random",
    private val game: Int = 4,
    timeout: Int = 60,
    ipAddress: String? = "localhost"
) : TablutClient(player, name, timeout, ipAddress) {

    override fun run() {
        try {
            declareName()
        } catch (e: Exception) {
            e.printStackTrace()
        }
        var state: State?
        var rules: Game? = null
        when (game) {
            1 -> {
                state = StateTablut()
                rules = GameTablut()
            }
            2 -> {
                state = StateTablut()
                rules = GameModernTablut()
            }
            3 -> {
                state = StateBrandub()
                rules = GameTablut()
            }
            4 -> {
                state = StateTablut()
                state.turn=Turn.WHITE
                rules = GameAshtonTablut(99, 0, "garbage", "fake", "fake")
                println("Ashton Tablut game")
            }
            else -> {
                println("Error in game selection")
                System.exit(4)
            }
        }
        val pawns: MutableList<IntArray> = ArrayList()
        val empty: MutableList<IntArray> = ArrayList()
        println("You are player " + player.toString() + "!")
        while (true) {
            try {
                read()
            } catch (e1: IOException) {
                // TODO Auto-generated catch block
                e1.printStackTrace()
                System.exit(1)
            }
            println("Current state:")
            state = currentState
            println(state.toString())
            try {
                Thread.sleep(1000)
            } catch (e: InterruptedException) {
                throw RuntimeException(e)
            }
            if (player == Turn.WHITE) {
                // Mio turno
                if (currentState?.turn == Turn.WHITE) {
                    var buf: IntArray
                    if (state != null) {
                        for (i in state.board?.indices!!) {
                            for (j in state.board!!.indices) {
                                if (state.getPawn(i, j)!!.equalsPawn(Pawn.WHITE.toString())
                                    || state.getPawn(i, j)!!.equalsPawn(Pawn.KING.toString())
                                ) {
                                    buf = IntArray(2)
                                    buf[0] = i
                                    buf[1] = j
                                    pawns.add(buf)
                                } else if (state.getPawn(i, j)!!.equalsPawn(Pawn.EMPTY.toString())) {
                                    buf = IntArray(2)
                                    buf[0] = i
                                    buf[1] = j
                                    empty.add(buf)
                                }
                            }
                        }
                    }
                    var selected: IntArray
                    var found = false
                    var a: Action? = null
                    a = Action("z0", "z0", Turn.WHITE)
                    while (!found) {
                        selected = if (pawns.size > 1) {
                            pawns[Random().nextInt(pawns.size - 1)]
                        } else {
                            pawns[0]
                        }
                        val from = currentState!!.getBox(selected[0], selected[1])
                        selected = empty[Random().nextInt(empty.size - 1)]
                        val to = currentState!!.getBox(selected[0], selected[1])
                        a = Action(from, to, Turn.WHITE)
                        found = try {
                            rules!!.checkMove(state, a)
                            true
                        } catch (e: Exception) {
                            throw RuntimeException(e)
                        }
                    }
                    println("Mossa scelta: " + Objects.requireNonNull(a))
                    try {
                        write(a)
                    } catch (e: IOException) {
                        // TODO Auto-generated catch block
                        e.printStackTrace()
                    }
                    pawns.clear()
                    empty.clear()
                } else if (state?.turn == Turn.BLACK) {
                    println("Waiting for your opponent move... ")
                } else if (state?.turn == Turn.WHITEWIN) {
                    println("YOU WIN!")
                    System.exit(0)
                } else if (state?.turn == Turn.BLACKWIN) {
                    println("YOU LOSE!")
                    System.exit(0)
                } else if (state?.turn == Turn.DRAW) {
                    println("DRAW!")
                    System.exit(0)
                }
            } else {

                // Mio turno
                if (currentState?.turn == Turn.BLACK) {
                    var buf: IntArray
                    if (state != null) {
                        for (i in state.board?.indices!!) {
                            for (j in state.board!!.indices) {
                                if (state.getPawn(i, j)!!.equalsPawn(Pawn.BLACK.toString())) {
                                    buf = IntArray(2)
                                    buf[0] = i
                                    buf[1] = j
                                    pawns.add(buf)
                                } else if (state.getPawn(i, j)!!.equalsPawn(Pawn.EMPTY.toString())) {
                                    buf = IntArray(2)
                                    buf[0] = i
                                    buf[1] = j
                                    empty.add(buf)
                                }
                            }
                        }
                    }
                    var selected: IntArray
                    var found = false
                    var a: Action? = null
                    a = Action("z0", "z0", Turn.BLACK)
                    while (!found) {
                        selected = pawns[Random().nextInt(pawns.size - 1)]
                        val from = currentState!!.getBox(selected[0], selected[1])
                        selected = empty[Random().nextInt(empty.size - 1)]
                        val to = currentState!!.getBox(selected[0], selected[1])
                        a = Action(from, to, Turn.BLACK)
                        println("try: " + Objects.requireNonNull(a))
                        found = try {
                            rules!!.checkMove(state, a)
                            true
                        } catch (e: Exception) {
                            throw RuntimeException(e)
                        }
                    }
                    println("Mossa scelta: $a")
                    try {
                        write(a)
                    } catch (e: IOException) {
                        // TODO Auto-generated catch block
                        e.printStackTrace()
                    }
                    pawns.clear()
                    empty.clear()
                } else if (state?.turn == Turn.WHITE) {
                    println("Waiting for your opponent move... ")
                } else if (state?.turn == Turn.WHITEWIN) {
                    println("YOU LOSE!")
                    System.exit(0)
                } else if (state?.turn == Turn.BLACKWIN) {
                    println("YOU WIN!")
                    System.exit(0)
                } else if (state?.turn == Turn.DRAW) {
                    println("DRAW!")
                    System.exit(0)
                }
            }
        }
    }

    companion object {
        @Throws(IOException::class)
        @JvmStatic
        fun main(args: Array<String>) {
            val gametype = 4
            var role = ""
            val name = "random"
            var ipAddress = "localhost"
            var timeout = 60
            // TODO: change the behavior?
            if (args.size < 1) {
                println("You must specify which player you are (WHITE or BLACK)")
                System.exit(-1)
            } else {
                println(args[0])
                role = args[0]
            }
            if (args.size == 2) {
                println(args[1])
                timeout = args[1].toInt()
            }
            if (args.size == 3) {
                ipAddress = args[2]
            }
            println("Selected client: " + args[0])
            val client = TablutRandomClient(role, name, gametype, timeout, ipAddress)
            client.run()
        }
    }
}